### DISCUSSION TITLE
french_downcase_string

### DISCUSSION THREAD

**Discussion by ddavout** [ddavout/FESTIfr#19] [Ideas]
la procédure french_downcase_string
```
(french_downcase_string name)
    Downcase a word and output it as a string
```
devrait, pour un peu plus de cohérence, s'appeler INST_LANG_downcase,
mais surtout 
serait plus _sexy_ si on en employait 

`tr [:upper:] [:lower:]`

**Comment by ddavout** [comment]
pour la cohérence, je suis d'accord.
que la procédure soit _artisanale_, je le reconnais.
**mais**
l'usage de tr n'est pas _recommandable_ :

`echo "Çà" | tr [:upper:] [:lower:] > ici_çà; grep 'Ç' ici_çà `

On ne change pas pour  une équipe qui perd !

**Comment by ddavout** [comment]
On en discute ? 
(essai de _Convert to discussion_)

**Comment by Pallas1303** [comment]
```scheme
(define (french_downcase_string name)
            (set! tmpfile (make_tmp_filename))
            (format t "%s\n" name) ;; Debug
            (set! tr_word (format nil "echo %s | tr [:upper:] [:lower:] > %s" name tmpfile))
            (system tr_word)


          (let ((fd (fopen tmpfile  "r")))
              (set! dword (readfp fd))
              (fclose fd)
          )
           (format t "%s\n" dword) ;; Debug
           (delete-file tmpfile)                              
                         
)
```

code non testé.

**Comment by ddavout** [comment]
### tr doesn't work with multi-byte characters
```
name="£ÙÀÇ"; echo "${name}" |  tr [:upper:] [:lower:]
£ùàç
```

### sed is a possibility
but I read that you should have a locale in utf8
### perl could be a solution
### bash using parameter expansion
if  it is not too old :
```
name="£ùàç"; echo "${name@U}"
£ÙÀÇ
```

but you have to make the procedure **system** understand bash.

**Reply by Pallas1303** [reply]
I understood, I found this:
https://unix.stackexchange.com/questions/228558/how-to-make-tr-aware-of-non-asciiunicode-characters

**Reply by Pallas1303** [reply]
```
name="£ùàç"; echo "${name@U}"
£ÙÀÇ
```

This seems to be functional.

**Reply by ddavout** [reply]
Yes, it is :)
but I'm worried about portability : it's a **runtime** procedure
I have no success to make **tr** working on this computer and it's the same problem : do we need some specific locales we don't have.

**Reply by ddavout** [reply]
I have more success with sed, but imagine that I don't have  the right `locale` installed, it's doesn't work

```
export LANG=US_fr..utf-8 ; echo $name | sed 's/[[:upper:]]/\l&/g'
£ÙÀÇ
```

**Reply by Pallas1303** [reply]
portability for Flite? 
For instalacion and configure of locale I used this:

```bash
sudo dpkg-reconfigure locales 
```

And set languagens for instalacion.

**Comment by ddavout** [comment]
Thanks,   I did reconfigure my locales,  and yet no success;
I give up     `tr`  
If I have to change my downcasing stuff, I may switch to    `awk`
`$ echo "ế1éêç" | awk '{print toupper($0)}`

**Reply by Pallas1303** [reply]
In link, the tr command not support this type de char. For set/force a lang in executed of a command.

Exemple
```bash
LANG=fr_FR.UTF-8 here_the_command
```

**Comment by ddavout** [comment]
How can I be more careful ?!
```
#!/usr/bin/env bash
(
  name="A£ÙÀÇµ"
  LANG=fr_FR.UTF-8
  LC_COLLATE=$LANG # to override any value it had before that would take precedence
  export LANG;
  export LC_COLLATE
  echo "$name" | tr '[:upper:]' '[:lower:]';
)
```
